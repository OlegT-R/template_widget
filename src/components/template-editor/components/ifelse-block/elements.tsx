import { Button } from 'antd';
import * as styledComponents from 'styled-components';

const { default: styled } = styledComponents;

export const Block = styled.div`
  width: 100%;
`;

export const BlockLine = styled.div`
  width: 100%;
  padding: 5px 0;
  display: flex;
  align-items: flex-start;
`;

export const OperatorCell = styled.div`
  width: 150px;
`;

export const ContentCell = styled.div`
  width: calc(100% - 150px);
`;

export const StyledBtn = styled(Button)`
  margin-left: 15px;
`;
