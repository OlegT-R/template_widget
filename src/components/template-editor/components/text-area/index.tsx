import React, { Component } from 'react';

import { Input } from 'antd';
import {
  ITextArea,
  NodeType
} from '../../../../models/template-model/entities';
import { SetActiveTextAreaType } from '../../utils/hooks';

const { TextArea: Area } = Input;

interface ITextAreaProps {
  node: ITextArea;
  parentNode: NodeType;
  onSetActiveTextArea: SetActiveTextAreaType;
}

interface IState {
  value: string;
  propsValue: string;
}

export class TextArea extends Component<ITextAreaProps, IState> {
  public static getDerivedStateFromProps(props: ITextAreaProps, state: IState) {
    if (props.node.value !== state.propsValue) {
      return {
        value: props.node.value,
        propsValue: props.node.value,
      };
    }

    return null;
  }

  public state = {
    value: '',
    propsValue: '',
  };

  public render() {
    const { node, onSetActiveTextArea, parentNode } = this.props;
    const { value } = this.state;
    return (
      <Area
        value={value}
        autosize={{ minRows: 2 }}
        onChange={e => {
          this.setState({ value: e.target.value });
        }}
        onBlur={e =>
          onSetActiveTextArea({
            node,
            parent: parentNode,
            position: e.target.selectionEnd,
            value: e.target.value,
          })
        }
      />
    );
  }
}
