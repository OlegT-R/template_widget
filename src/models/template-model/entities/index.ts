export * from './node';
export * from './template';
export * from './textarea';
export * from './variables';
