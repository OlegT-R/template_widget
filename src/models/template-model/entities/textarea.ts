export interface ITextArea {
    id: string;
    value: string;
}
