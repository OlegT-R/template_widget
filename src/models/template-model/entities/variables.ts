
export type VariableNamesType = string[];

export interface IVariables {
  [key: string]: string,
}
