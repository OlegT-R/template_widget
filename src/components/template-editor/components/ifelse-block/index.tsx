import React from 'react';

import { Icon } from 'antd';
import { IIfElse, NodeType } from '../../../../models/template-model/entities';

import { DeleteNodeType, SetActiveTextAreaType } from '../../utils/hooks';
import { getTemplateNode } from '../template-node-switcher';

import { Operator } from '../ifelse-btn/elements';

import {
  Block,
  BlockLine,
  ContentCell,
  OperatorCell,
  StyledBtn
} from './elements';

interface IIfElseBlockProps {
  onDelete: DeleteNodeType;
  node: IIfElse;
  parentNode: NodeType;
  onSetActiveTextArea: SetActiveTextAreaType;
}

export const IfElseBLock: React.FC<IIfElseBlockProps> = ({
  node,
  onDelete,
  parentNode,
  onSetActiveTextArea,
}) => (
  <Block>
    <BlockLine>
      <OperatorCell>
        <Operator>If</Operator>
        <StyledBtn
          type='danger'
          size='small'
          onClick={() => onDelete({ node, parentNode})}
        >
          delete
          <Icon type='delete' />
        </StyledBtn>
      </OperatorCell>
      <ContentCell>
        {getTemplateNode(node.if, onDelete, onSetActiveTextArea)}
      </ContentCell>
    </BlockLine>
    <BlockLine>
      <OperatorCell>
        <Operator>Then</Operator>
      </OperatorCell>
      <ContentCell>
        {getTemplateNode(node.then, onDelete, onSetActiveTextArea)}
      </ContentCell>
    </BlockLine>
    <BlockLine>
      <OperatorCell>
        <Operator>Else</Operator>
      </OperatorCell>
      <ContentCell>
        {getTemplateNode(node.else, onDelete, onSetActiveTextArea)}
      </ContentCell>
    </BlockLine>
  </Block>
);
