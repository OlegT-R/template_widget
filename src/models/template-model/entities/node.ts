import { ITextArea } from './textarea';

export interface IIfElse {
  id: string,
  if: NodeType;
  then: NodeType;
  else: NodeType;
}

export type NodeType = Array<ITextArea | IIfElse>;
