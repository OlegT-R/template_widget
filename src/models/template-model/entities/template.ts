import { IIfElse, NodeType } from './node';
import { ITextArea } from './textarea';

export interface ITemplate {
  block: NodeType;
}

export type NodeItem = ITextArea | IIfElse;

export function isTextArea(node: NodeItem): node is ITextArea {
  return (node as ITextArea).value !== undefined;
}

export function isIfElse(node: NodeItem): node is IIfElse {
  return (node as IIfElse).if !== undefined;
}
