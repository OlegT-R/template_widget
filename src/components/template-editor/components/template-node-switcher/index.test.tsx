import React from 'react';

import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { NodeType } from '../../../../models/template-model/entities';
import { IfElseBLock } from '../ifelse-block';
import { TextArea } from '../text-area';

import { getTemplateNode } from '.';

const onDeleteNode = () => null;
const onSetActiveTextArea = () => null;
describe('<MyComponent />', () => {
  it('should render <TextArea />', () => {
    const textAreaNode = {
      id: 'testTextArea',
      value: 'TextArea',
    };

    const wrapper = shallow(
      <div>
        {getTemplateNode([textAreaNode], onDeleteNode, onSetActiveTextArea)}
      </div>
    );
    expect(wrapper.find(TextArea).length).toBe(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render <IfElseBLock />', () => {
    const node: NodeType = [
      {
        id: 'id',
        if: [
          {
            id: 'testifElseIF',
            value: 'testifElseIF',
          },
        ],
        then: [
          {
            id: 'testifElseTHEN',
            value: 'testifElseTHEN',
          },
        ],
        else: [
          {
            id: 'testifElseELSE',
            value: 'testifElseELSE',
          },
        ],
      },
    ];

    const wrapper = shallow(
      <div>{getTemplateNode(node, onDeleteNode, onSetActiveTextArea)}</div>
    );
    expect(wrapper.find(IfElseBLock).length).toBe(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
