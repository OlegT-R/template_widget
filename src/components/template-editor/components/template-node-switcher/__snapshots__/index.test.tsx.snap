// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MyComponent /> should render <IfElseBLock /> 1`] = `
<div>
  <IfElseBLock
    key="id"
    node={
      Object {
        "else": Array [
          Object {
            "id": "testifElseELSE",
            "value": "testifElseELSE",
          },
        ],
        "id": "id",
        "if": Array [
          Object {
            "id": "testifElseIF",
            "value": "testifElseIF",
          },
        ],
        "then": Array [
          Object {
            "id": "testifElseTHEN",
            "value": "testifElseTHEN",
          },
        ],
      }
    }
    onDelete={[Function]}
    onSetActiveTextArea={[Function]}
    parentNode={
      Array [
        Object {
          "else": Array [
            Object {
              "id": "testifElseELSE",
              "value": "testifElseELSE",
            },
          ],
          "id": "id",
          "if": Array [
            Object {
              "id": "testifElseIF",
              "value": "testifElseIF",
            },
          ],
          "then": Array [
            Object {
              "id": "testifElseTHEN",
              "value": "testifElseTHEN",
            },
          ],
        },
      ]
    }
  />
</div>
`;

exports[`<MyComponent /> should render <TextArea /> 1`] = `
<div>
  <TextArea
    key="testTextArea"
    node={
      Object {
        "id": "testTextArea",
        "value": "TextArea",
      }
    }
    onSetActiveTextArea={[Function]}
    parentNode={
      Array [
        Object {
          "id": "testTextArea",
          "value": "TextArea",
        },
      ]
    }
  />
</div>
`;
